KAFKA-15514: Metadata records Replicas->Assignment

soarez opened the pull request:
The new "Assignments" field replaces the "Replicas" field in PartitionRecord and PartitionChangeRecord.

Depends on #14290 - [KAFKA-15355](https://issues.apache.org/jira/browse/KAFKA-15355)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soarez committed:
KAFKA-15514: Metadata records Replicas->Assignment

The new "Assignments" field replaces the "Replicas"
field in PartitionRecord and PartitionChangeRecord.

soarez committed:
KAFKA-15355: Message schema changes

https://issues.apache.org/jira/browse/KAFKA-15355

OmniaGM commented on core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala:
```diff
@@ -63,6 +63,7 @@ class KRaftMetadataCache(val brokerId: Int) extends MetadataCache with Logging w
                                        brokers: Array[Int],
                                        listenerName: ListenerName,
                                        filterUnavailableEndpoints: Boolean): java.util.List[Integer] = {
+    // TODO KAFKA-15362
```
Do we know what the perf impact on metadata requests is by adding the logic of KAFKA-15362 here? Especially with the warning comment in [L60](https://github.com/apache/kafka/blob/4612fe42af0df0a4c1affaf66c55d01eb6267ce3/core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala#L60) regard adding extra logic to `KRaftMetadataCache.maybeFilterAliveReplicas`.

OmniaGM commented on core/src/main/scala/kafka/zk/migration/ZkTopicMigrationClient.scala:
```diff
@@ -102,7 +111,7 @@ class ZkTopicMigrationClient(zkClient: KafkaZkClient) extends TopicMigrationClie
     val assignments = partitions.asScala.map { case (partitionId, partition) =>
       new TopicPartition(topicName, partitionId) ->
-        ReplicaAssignment(partition.replicas, partition.addingReplicas, partition.removingReplicas)
+        ReplicaAssignment(partition.replicaBrokerIds(), partition.addingReplicas, partition.removingReplicas)
```
scala nit: It is usually recommended to not use empty parentheses in scala when a parameterless function has no side effect in both the definition of the function and when they call it (which is the case for `partition.replicaBrokerIds`).

OmniaGM commented on core/src/test/scala/integration/kafka/zk/ZkMigrationIntegrationTest.scala:
```diff
@@ -128,7 +128,7 @@ class ZkMigrationIntegrationTest {
     val underlying = clusterInstance.asInstanceOf[ZkClusterInstance].getUnderlying()
     val zkClient = underlying.zkClient
-    val migrationClient = ZkMigrationClient(zkClient, PasswordEncoder.noop())
+    val migrationClient = ZkMigrationClient(zkClient, PasswordEncoder.noop(), MetadataVersion.latest())
```
nit: Same as `replicaBrokerIds`, `MetadataVersion.latest` has no side-effect so it should be called without empty parentheses.

OmniaGM commented:
I believe this test is failing due to some changes here https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14516/3/testReport/junit/kafka.zk.migration/ZkMigrationClientTest/Build___JDK_21_and_Scala_2_13___testUpdateExistingTopicWithNewAndChangedPartitions__/ .

soarez commented:
Thanks for the review and comments @OmniaGM . Depending on the outcome of the current discussion in #14290 there may be some bigger changes to this PR, so I'll wait until that's resolved before addressing your comments.

soarez commented:
Closing this due to the change of plan in  #14290
